### DISCUSSION TITLE
Accessing Prior Variants/Submissions of A Question When Grading

### DISCUSSION THREAD

**Discussion by mfbutner** [PrairieLearn/PrairieLearn#7570] [General]
Right now I'm using the external grader for grading questions where I ask students to write programs. Most of the inputs for the test cases are randomly generated. Sometimes a student will submit and fail a few test cases and then just resubmit their program again with no changes and pass all of the test cases. What I'd like to be able to do is keep track of all of the tests that failed in all prior submissions so as to be able to rerun those explicit test cases in addition to any new ones whenever the student submits again. 

Is it possible to do what I want with how things are now? If not, do you think something like this will ever be added to Prairie Learn or is it just too niche? If it is too niche, any suggestions on how to avoid my original problem?

**Comment by eliotwrobson** [comment]
I think to do something like this, the external grader would need to have the ability to read and write from `data['params']` or similar. I think that ability seems pretty niche, but might not be an unreasonable thing to add? However, I think the solution to this problem is to find a way to rely less on random generation.

**Reply by mwest1066** [reply]
We've talked before about giving the external grader equivalent functionality to the internal grader (basically making them identical in capabilities, just run differently). This would include modifying `params` in the same way that the internal grader currently can.

**Comment by mwest1066** [comment]
We don't currently have a way of passing data like this back from the external grader, but I think that it will be possible once we allow the external grader to modify the `params` data (which would allow it to save the test case data that it used). This is a planned change on the road map.

Until we add this, one work-around might be to make the set of random test cases use a fixed random seed per student, so that they'll always get the same set of test cases (but different to all other students). One way to do this would be to generate `params["random_seed"]` randomly for each student, and use this to initialize the random number generator in the external grader.

Alternatively, another possibility is to just run more random test cases. In the case you describe it sounds like the student code is incorrect but the test cases aren't catching it? More test cases (whether random or hard-coded) feels like one approach to detect this.

**Reply by nwalters512** [reply]
I was typing out almost exactly this response before you beat me to it!

Instead of `params["random_seed"]`, I'd just use `variant_seed` from `data.json` in the grader. That's going to be both unique and stable per variant, which is I think exactly what we want here.

+1 on adding more test cases. "bad code gets 100%" feels like something that can and should be solved on its own.

**Comment by mfbutner** [comment]
One issue I've had in the past of "fixing" the test case input is that students will write their solution to just match the output. They do something like
- If the input is test case 1, print answer 1
- If the input is test case 2, print answer 2
- ...

So I'm trying to proactively avoid this by using randomization. I have also played around with doing hidden test cases but those sometimes lead to a lot of frustration for the student if they can see they are failing but don't know what they are failing on and while the visible test cases are supposed to be equivalent, sometimes something that is supposed to be equivalent is somehow different enough to trigger their bug.

**Reply by jonatanschroeder** [reply]
I usually do that by combining fixed test cases (which help students debug their code) and random test cases (which take care of hardcoding). This also allows me to review cases where the fixed cases passed but the random cases failed, which is usually a good indication that some level of hardcoding took place. Also, I have a lot of test cases.

**Reply by jonatanschroeder** [reply]
Fixed test cases also help with things like corner cases, which are harder to pick with random cases.

**Reply by mfbutner** [reply]
I do have fixed test cases for edge/special cases and some normal test cases but occasionally there is some random test case that conceptually isn't any different from a fixed test case and the student's code breaks on it. That's the thing I'm not quite sure how to address.
